File: exercises/101-109_small_problems/easy_01/easy_06.rb
# easy_06.rb - Reverse It (Part 2)

def reverse_words(string)
  arr = string.split
  arr.each { |sub_string| sub_string.reverse! if sub_string.length >= 5 }
  arr.join(' ')
end

=begin
utp
- take a string of 1 or more words
- reverse words whose length => 5 characters
- return the string with spaces only if more than one word present

i/o/d
- i: string
- o: string
- d: array

test cases:
puts reverse_words('Professional')
puts reverse_words('Walk around the block')
puts reverse_words('Launch School')

algorithm
- split string and iterate over each word; if word =>5 then reverse it, then
  join array as string with ' ' as delimiter

=end

File: exercises/101-109_small_problems/easy_01/go/easy_06.go
package main

import (
	"fmt"
	"strings"
)

func reverseWords(s string) string {
	words := strings.Split(s, " ")
	reversed := []string{}

	for _, word := range words {
		if len([]rune(word)) > 4 {
			runes := []rune(word)

			for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
				runes[i], runes[j] = runes[j], runes[i]
			}

			word = string(runes)
		}

		reversed = append(reversed, word)
	}

	return strings.Join(reversed, " ")
}

func main() {
	fmt.Println(reverseWords("Professional"))
	fmt.Println(reverseWords("Walk around the block"))
	fmt.Println(reverseWords("Launch School"))
}
